### ISSUE TITLE
Invalid divisor in softmax layer

### ISSUE THREAD

**Issue by ruamzuzla** [Dobiasd/frugally-deep#69]
Hello,

I ran into the following issue:

> terminate called after throwing an instance of 'std::runtime_error'
  what():  Invalid divisor in softmax layer. Try using
#define FDEEP_FLOAT_TYPE double
before
#include "fdeep/fdeep.hpp"
If you receive this error while using fdeep::load_model(filepath) try fdeep::load_model(filepath, false) instead to skip the verification step.

My model is working fine with keras. Therefore I created this SSCCE:

```c++
#define FDEEP_FLOAT_TYPE double
#include <fdeep/fdeep.hpp>
#include <opencv2/core/core.hpp>
#include <iostream>

int main()
{
    const auto model = fdeep::load_model("fdeep_model.json", false);

    cv::Mat inputImage = cv::Mat(150, 75, CV_8UC3, cv::Scalar(255, 0, 0));

    const auto input = fdeep::tensor3_from_bytes(inputImage.ptr(),
						 inputImage.rows,
						 inputImage.cols,
						 inputImage.channels(),
						 0.0f,
						 255.0f);

    std::cout << fdeep::show_shape3s(model.get_input_shapes()) << std::endl;
    std::cout << fdeep::show_shape3s({input.shape()}) << std::endl;

    auto pred = model.predict({input});
    std::cout << fdeep::show_tensor3s(pred) << std::endl;
}
```

As you can see, I already tried the suggestions in the error message. Strangely, I observed that the prediction does work, when the image is initialized completely black:

```c++
cv::Mat inputImage = cv::Mat(150, 75, CV_8UC3, cv::Scalar(0, 0, 0));
```

and it also works with some other values like:

```c++
cv::Mat inputImage = cv::Mat(150, 75, CV_8UC3, cv::Scalar(1, 2, 3));
```

This is really confusing to me and I'm not sure how to proceed from here. Do you have any suggestions for me?

I uploaded my .h5 here: https://drive.google.com/open?id=1g4jvHllyTtUZ0n_Yz-wJqkeeUHfafWpy
My OpenCV version is 3.3.1.

Thanks for your help!

**Comment by Dobiasd** [comment]
First of all thanks for this very good style of reporting the issue.

I was able to reproduce the error on my system and will start to investigate now.

**Comment by Dobiasd** [comment]
Loading the model seems to not be the problem. So I can use

```cpp
const auto model = fdeep::load_model("issue69/fdeep_model.json");
```
instead of
```cpp
const auto model = fdeep::load_model("issue69/fdeep_model.json", false);
```

The prediction step however really crashes in this code:

```cpp
#define FDEEP_FLOAT_TYPE double
#include <fdeep/fdeep.hpp>
#include <opencv2/core/core.hpp>
#include <iostream>
int main()
{
    const auto model = fdeep::load_model("issue69/fdeep_model.json");
    cv::Mat inputImage = cv::Mat(150, 75, CV_8UC3, cv::Scalar(255, 0, 0));
    const auto input = fdeep::tensor3_from_bytes(inputImage.ptr(),
        static_cast<std::size_t>(inputImage.rows),
        static_cast<std::size_t>(inputImage.cols),
        static_cast<std::size_t>(inputImage.channels()),
        0.0f, 255.0f);
    const auto pred = model.predict({input});
}
```

But when I change the call to `fdeep::tensor3_from_bytes` to scale into the range `[0.0, 1.0]` instead of `[0.0, 255.0]`, it works OK:

```cpp
#define FDEEP_FLOAT_TYPE double
#include <fdeep/fdeep.hpp>
#include <opencv2/core/core.hpp>
#include <iostream>
int main()
{
    const auto model = fdeep::load_model("issue69/fdeep_model.json");
    cv::Mat inputImage = cv::Mat(150, 75, CV_8UC3, cv::Scalar(255, 0, 0));
    const auto input = fdeep::tensor3_from_bytes(inputImage.ptr(),
        static_cast<std::size_t>(inputImage.rows),
        static_cast<std::size_t>(inputImage.cols),
        static_cast<std::size_t>(inputImage.channels()),
        0.0f, 1.0f);
    const auto pred = model.predict({input});
}
```

So my first question is: Do you really want your floating point image pixels (i.e. the values of the input tensor to the model) to be in the range of `[0.0, 255.0]` instead of `[0.0, 1.0]`? I'm asking because this is quite unusual in my experience.

**Comment by ruamzuzla** [comment]
Thank you for the fast response! My model was trained on a numpy byte array, which contains only the values 0 and 255 for different channels. I didn't apply any normalization, so I believe 255 is right, isn't it?

Also, I was using a regression model prior to this with very similar data and it worked alright.

**Comment by Dobiasd** [comment]
Yes, in that case you are right to also use the [0.0, 255.0] interval in C++.

I know this might not be the most elegant solution, but perhaps it's a quick workaround: Can you re-train your model in Keras with values in [0.0, 1.0], re-export to fdeep and try again?

**Comment by ruamzuzla** [comment]
Sure, I'm going to that right now. I will report back soon (or maybe tomorrow)

**Comment by ruamzuzla** [comment]
It worked! Thank you
